File: tests/fixtures/sample.rs
/**
 * Sample Rust file for testing
 * Contains structs, functions, traits, and implementations
 */

use std::collections::HashMap;
use std::io::{self, Read, Write};

// Type alias
pub type UserId = u64;
pub type Result<T> = std::result::Result<T, Box<dyn std::error::Error>>;

/// User struct with authentication details
#[derive(Debug, Clone)]
pub struct User {
    pub id: UserId,
    pub email: String,
    pub password_hash: String,
    pub role: UserRole,
}

/// User role enumeration
#[derive(Debug, Clone, PartialEq)]
pub enum UserRole {
    Admin,
    Moderator,
    User,
}

/// Authentication service trait
pub trait AuthService {
    fn login(&self, email: &str, password: &str) -> Result<User>;
    fn create_session(&self, user_id: UserId) -> Result<String>;
    fn verify_password(&self, password: &str, hash: &str) -> bool;
}

/// Default authentication service implementation
pub struct DefaultAuthService {
    users: HashMap<String, User>,
    session_timeout: u64,
}

impl DefaultAuthService {
    /// Create a new authentication service
    ///
    /// # Arguments
    /// * `session_timeout` - Session timeout in seconds
    ///
    /// # Returns
    /// A new `DefaultAuthService` instance
    pub fn new(session_timeout: u64) -> Self {
        DefaultAuthService {
            users: HashMap::new(),
            session_timeout,
        }
    }

    /// Generate a random session ID
    fn generate_session_id(&self) -> String {
        use rand::Rng;
        let mut rng = rand::thread_rng();

        (0..32)
            .map(|_| {
                let idx = rng.gen_range(0..62);
                match idx {
                    0..=9 => (b'0' + idx) as char,
                    10..=35 => (b'a' + idx - 10) as char,
                    _ => (b'A' + idx - 36) as char,
                }
            })
            .collect()
    }

    /// Add a user to the service
    pub fn add_user(&mut self, user: User) {
        self.users.insert(user.email.clone(), user);
    }
}

impl AuthService for DefaultAuthService {
    /// Authenticate user with email and password
    fn login(&self, email: &str, password: &str) -> Result<User> {
        if email.is_empty() || password.is_empty() {
            return Err("Email and password are required".into());
        }

        let user = self.users.get(email)
            .ok_or("User not found")?;

        if !self.verify_password(password, &user.password_hash) {
            return Err("Invalid password".into());
        }

        Ok(user.clone())
    }

    /// Create a new session for the user
    fn create_session(&self, user_id: UserId) -> Result<String> {
        let session_id = self.generate_session_id();

        // In a real implementation, store session in database
        println!("Created session {} for user {}", session_id, user_id);

        Ok(session_id)
    }

    /// Verify password against hash
    fn verify_password(&self, password: &str, hash: &str) -> bool {
        // Simplified for testing - in production use bcrypt or argon2
        password == hash
    }
}

/// Check if user has required permission
///
/// # Arguments
/// * `user` - The user to check
/// * `required_role` - The required role
///
/// # Returns
/// `true` if user has sufficient permissions
pub fn has_permission(user: &User, required_role: &UserRole) -> bool {
    let role_hierarchy = |role: &UserRole| -> u8 {
        match role {
            UserRole::Admin => 3,
            UserRole::Moderator => 2,
            UserRole::User => 1,
        }
    };

    role_hierarchy(&user.role) >= role_hierarchy(required_role)
}

/// Calculate cyclomatic complexity metrics for demonstration
///
/// # Arguments
/// * `data` - Input data array
///
/// # Returns
/// A tuple of (mean, max) values
pub fn calculate_complexity(data: &[i32]) -> (f64, i32) {
    if data.is_empty() {
        return (0.0, 0);
    }

    let mut total = 0i64;
    let mut max_value = data[0];

    // Calculate sum and max (demonstrates loops and conditionals)
    for &value in data.iter() {
        total += value as i64;

        if value > max_value {
            max_value = value;
        }
    }

    let mean = total as f64 / data.len() as f64;

    (mean, max_value)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_has_permission() {
        let admin_user = User {
            id: 1,
            email: "admin@test.com".to_string(),
            password_hash: "hash".to_string(),
            role: UserRole::Admin,
        };

        assert!(has_permission(&admin_user, &UserRole::User));
        assert!(has_permission(&admin_user, &UserRole::Moderator));
        assert!(has_permission(&admin_user, &UserRole::Admin));
    }

    #[test]
    fn test_calculate_complexity() {
        let data = vec![1, 2, 3, 4, 5];
        let (mean, max) = calculate_complexity(&data);

        assert_eq!(mean, 3.0);
        assert_eq!(max, 5);
    }
}

File: tests/fixtures/sample.go
// Package auth provides user authentication services
package auth

import (
	"database/sql"
	"errors"
	"fmt"
	"time"
)

// SessionTimeout defines the session expiration time
const SessionTimeout = 3600 // 1 hour in seconds

// APIVersion defines the current API version
const APIVersion = "v1"

// User represents a user in the system
type User struct {
	ID           string
	Email        string
	PasswordHash string
	Role         UserRole
}

// UserRole represents user permission levels
type UserRole int

const (
	RoleUser UserRole = iota
	RoleModerator
	RoleAdmin
)

// AuthService handles user authentication
type AuthService struct {
	dbClient *sql.DB
	timeout  int
}

// NewAuthService creates a new authentication service
func NewAuthService(dbClient *sql.DB) *AuthService {
	return &AuthService{
		dbClient: dbClient,
		timeout:  SessionTimeout,
	}
}

// Login authenticates a user with credentials
// Returns the user if authentication succeeds, error otherwise
func (s *AuthService) Login(email, password string) (*User, error) {
	if email == "" || password == "" {
		return nil, errors.New("email and password are required")
	}

	user, err := s.queryUser(email)
	if err != nil {
		return nil, err
	}

	if user == nil {
		return nil, errors.New("user not found")
	}

	isValid := s.verifyPassword(password, user.PasswordHash)
	if !isValid {
		return nil, errors.New("invalid password")
	}

	return user, nil
}

// queryUser fetches a user by email
func (s *AuthService) queryUser(email string) (*User, error) {
	var user User
	err := s.dbClient.QueryRow(
		"SELECT id, email, password_hash, role FROM users WHERE email = ?",
		email,
	).Scan(&user.ID, &user.Email, &user.PasswordHash, &user.Role)

	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	return &user, nil
}

// verifyPassword checks if password matches hash
func (s *AuthService) verifyPassword(password, hash string) bool {
	// Simplified for testing
	return password == hash
}

// CreateSession creates a new user session
func (s *AuthService) CreateSession(userID string) (string, error) {
	sessionID := generateSessionID()
	expiresAt := time.Now().Add(time.Duration(s.timeout) * time.Second)

	_, err := s.dbClient.Exec(
		"INSERT INTO sessions (id, user_id, expires_at) VALUES (?, ?, ?)",
		sessionID, userID, expiresAt,
	)

	if err != nil {
		return "", err
	}

	return sessionID, nil
}

// generateSessionID creates a random session ID
func generateSessionID() string {
	return fmt.Sprintf("session_%d", time.Now().UnixNano())
}

// PermissionChecker defines interface for permission checking
type PermissionChecker interface {
	HasPermission(user *User, requiredRole UserRole) bool
}

// HasPermission checks if user has required permission level
func HasPermission(user *User, requiredRole UserRole) bool {
	return user.Role >= requiredRole
}

// CalculateComplexity demonstrates control flow complexity
func CalculateComplexity(data []int) map[string]float64 {
	if len(data) == 0 {
		return map[string]float64{
			"mean": 0.0,
			"max":  0.0,
		}
	}

	total := 0
	for _, value := range data {
		total += value
	}

	mean := float64(total) / float64(len(data))

	max := data[0]
	for _, value := range data {
		if value > max {
			max = value
		}
	}

	return map[string]float64{
		"mean": mean,
		"max":  float64(max),
	}
}
